ASoC: SOF: ipc4-topology: Use correct queue_id for requesting input p…

ujfalusi opened the pull request:
…in format

It is incorrect to request the input pin format of the sink widget using the pin index of the output of the source module as the indexes are not necessarily matching.
moduleA.out_pin1 can be connected to moduleB.in_pin0 for example.

Use the dst_queue_id to request the input format of the sink module.

ujfalusi committed:
ASoC: SOF: ipc4-topology: Use correct queue_id for requesting input pin format

It is incorrect to request the input pin format of the destination widget
using the output pin index of the source module as the indexes are not
necessarily matching.
moduleA.out_pin1 can be connected to moduleB.in_pin0 for example.

Use the dst_queue_id to request the input format of the destination module.

This bug remained unnoticed likely because in nocodec topologies we don't
have process modules after a module copier, thus the pin/queue index is
ignored.
For the process module case, the code was likely have been tested in a
controlled way where all the pin/queue/format properties were present to
work.

Update the debug prints to have better information.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>
Cc: <stable@vger.kernel.org> # v6.8+

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2871,7 +2871,7 @@ static int sof_ipc4_set_copier_sink_format(struct snd_sof_dev *sdev,
 
 	dev_dbg(sdev->dev, "%s set copier sink %d format\n",
-		src_widget->widget->name, sink_id);
+		src_widget->widget->name, sroute->src_queue_id);
```
This is very confusing. Can you remind us in the commit message what the convention is for sinks/sources, having a log message that says 'sink N' where N is a source_queue_id is not really helping....

The gstreamer convention is that a sink receives data, but I think IPC4 provides data on a sink pin?

ujfalusi replied:
The connection in IPC4 is:
src_widget.pin[sroute->src_queue_id] -> sink_widget.pin[sroute->dst_queue_id]
Before this patch we used the same queue_id (or pin_id) to get the format from the sink_widget, which was incorrect.

plbossart replied:
@ujfalusi the problem I have is that the dev_dbg log mixes sink and source.

ujfalusi replied:
it prints things correctly.
We are setting the src_widget's sink (pin) format, that is src_widget.pin[sroute->src_queue_id]
We do that by copying the input format of the destination widget, and we say that we cannot get it if it is not found.

But right, this is amazingly confusing..
the source widget have sink pin, which is connected to a sink widget's source pin, we could refer to both as s_swidget / s_pin/s_queue all around to make things more fun :D

ujfalusi replied:
@plbossart, I think the right course of action is to remove this print as such. We will print the connection anyways and this brings no value, we will print on error.

ujfalusi commented:
Changes since v1:
- Debug prints slightly modified to be less confusing
- One print is dropped since it had no real value
- stable tag added for 6.8+ backport
- commit message updated

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2947,15 +2946,17 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
 	if (sroute->src_queue_id < 0) {
-		dev_err(sdev->dev, "failed to get queue ID for source widget: %s\n",
+		dev_err(sdev->dev,
+			"failed to get output queue ID for source widget: %s\n",
```
I would use the same terminology as the code

"failed to get src_queue_id ID for source widget:" 

our source queue ID as use line 2971

ujfalusi replied:
right, but it is the output pin/queue ID of the source widget.
In sroute we store it as the src_queue_id, yes, but when you say: "failed to get src_queue_id ID for source widget:" I would read it that it is the source queue ID of the source widget that failed, while it is not, it is the sink_queue_id of the source widget, which is a source queue in the connection between the two modules.

just overlooking the fact that we use `queue` and `pin` for the same thing, depending on where we are in code..

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2947,15 +2946,17 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
 	if (sroute->dst_queue_id < 0) {
-		dev_err(sdev->dev, "failed to get queue ID for sink widget: %s\n",
+		dev_err(sdev->dev,
+			"failed to get input queue ID for sink widget: %s\n",
```
I would use the same terminology as the code

"failed to get dst_queue_id ID for sink widget:"

ujfalusi replied:
here the dst_queue_id is the source queue of the sink widget.
I'm not sure which is better, but if we refer to the widget then it is from the widget's pow?

plbossart replied:
My point is that we don't need to add a new 'input' term that's not used anywhere else.
we only use 'input' and 'output' for formats.

ujfalusi replied:
"failed to get src_queue_id ID from source widget: %s"
and
"failed to get dst_queue_id ID from sink widget: %s"

?

ujfalusi commented:
Changes since v2:
- reworded the debug prints for src/dst queue ID request errors.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2972,8 +2973,8 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
+						      sink_widget, sroute);
 		if (ret < 0) {
 			dev_err(sdev->dev, "failed to set sink format for %s source queue ID %d\n",
 				src_widget->widget->name, sroute->src_queue_id);
```
maybe change this one as well for consistency?

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2869,7 +2869,7 @@ static void sof_ipc4_put_queue_id(struct snd_sof_widget *swidget, int queue_id,
 					   struct snd_sof_widget *src_widget,
 					   struct snd_sof_widget *sink_widget,
-					   int sink_id)
+					   struct snd_sof_route *sroute)
```
@ujfalusi I guess this worked before without issues only because we never had a process type widget connected to a module copier with multiple output pins before right? Maybe its worth mentioning it

ujfalusi replied:
@ranj063, likely that is the case, in nocodec topology we have gains after the module-copier branch.
When it comes to process module, I guess it was tested with a controlled module which happen to have multiple input pins, thus multiple formats per pins or the module copier on front of it had a single output, not sure, but I would rather avoid speculating in commit message.
Having said that, I will check if I can update it.

ujfalusi commented:
Changes since v3:
- commit message updated
- Error print for copier sink format setting failure updated also.

plbossart reviewed (approved):
thanks @ujfalusi

plbossart commented:
Humm, something's wrong but likely unrelated to this PR

https://sof-ci.01.org/linuxpr/PR5019/build3274/devicetest/index.html?model=MTLP_SDW_AIOC&testcase=check-suspend-resume-without-audio-20

Redoing the tests.

plbossart commented:
SOFCI TEST

ujfalusi commented:
@plbossart, same suspend-resume issue

ujfalusi commented:
SOFCI TEST

ujfalusi commented:
Doh, now `check-sof-logger.sh` failing across all devices...
